designs/0005-cmdstan3.md

```diff
@@ -0,0 +1,161 @@
+---
+output:
+  html_document: default
+  pdf_document: default
+---
+
+- *Feature Name:* CmdStan3
+- *Start Date:* 2020-01-16
+- *RFC PR(S):*
+- *Stan-dev Issue(s):*
+
+# CmdStan3 Summary
+[summary]: #summary
+
+CmdStan3 is a command-line interface to the Stan services layer
+which provides an extensible syntax for calling Stan services
+as well as model compilation.
+CmdStan3 provides a utility program `stan` which takes as arguments
+a set of subcommands; these subcommands map to functions
+exposed by the `stan::services` package, model compilation,
+and existing Stan utilities such as `stansummary`.
+
+
+# Motivation
+[motivation]: #motivation
+
+1. Unify all workflow operations: single utility for model compilation, fitting, and diagnostic operations.
+
+2. Usability: flat command syntax which allows configuration arguments to be specified in any order
+
+3. Extensibility: easy to add new commands corresponding to new services and to modify argument structure for existing commands.
+
+# Guide-level explanation: Command Syntax
+[guide-level-explanation]: #guide-level-explanation
+
+
+The first argument to the `stan` utility is either
+
+- `-h` or `--help` for help
+- `-v` or `--version` for CmdStan3 version information
+- a subcommand
+
+Subcommand arguments are then followed by either
+
+- `-h` or `--help` for subcommand-specific help
+- a Stan model name
+
+For subcommands which call a Stan service,
+the remaining command line arguments consist of an
+unordered set of pairs of the form `<name>=<value>`
+which are parsed and validated by the `stan` utility
+and then passed in to the services method.
+
+Possible exceptions to this subcommand syntax are:
+
+- The `compile` subcommand, which would instead take a series of makefile options.
+- Utilities such as `stansummary` which operate on one or more files but don't need a compiled model.
+
+
+### Examples
+
+- Print top-level help:
+```
+stan -h
+```
+
+- Print subcommand help:
+```
+stan optimize -h
+```
+
+- Compile a Stan program `my_model.stan` using CmdStan makefile _from any directory_:
+
+```
+stan compile my_model
+stan compile path/to/my_model
+stan compile path/to/my_model -makeflag1=value1 ... -makeflagN=valueN
+```
+
+
+- Fit a Stan program to data
+  + specify input, output, and sampling arguments in any order
+  + specify of non-default initial radius as well as parameter-specific initializations
+
+```
+stan sample my_model metric=dense data_file=my_data.json\
+            init_file=some_params.json init_radius=0.333 output_file=my_model.csv adapt_delta=0.95
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Interaction with other features:
+
+- CmdStan3 makes calls to GnuMake and therefore will spawn a process on the underlying OS; must be robust to path and filenames
```

Comment:
What is the plan for the language in which this utility program that will spawn other processes will be written? Is the plan to stay with C++?

Reply:
it would be a script - ideally python, but bash would be OK too.
in order to it to be Python, we need to have a robust installer for the toolchain.